include/sway/util.h

```diff
@@ -0,0 +1,14 @@
+#ifndef _SWAY_SWAY_UTIL_H
+#define _SWAY_SWAY_UTIL_H
+
+#include <spawn.h>
+
+/**
+ * Close fd and log theoretical case when close(2) failed
+ */
+void close_warn(int fd);
```

Comment:
Not a fan of this

Reply:
As I've already mentioned there is no point in checking return value of close()
So close_warn() can be just replaced by close()
Is it ok for you?

Reply:
No, we don't want to leak FDs.

Reply:
Leak fd in close?!
How checking fd result can help?

Reply:
If `close` fails, something is going very wrong. We should definitely hard fail.

Reply:
Hard fail means something like assert
Logging and 'failing' client spawn is not hard fail.
Also if you have not already, please, check links regarding close errors
But here, when we are working with sockets/pipes (not something on NFS) close can not fail...

Reply:
>Hard fail means something like assert

No. Assertions should be used if our internal state is inconsistent. That's not the case here.

We have many cases in which very bad things happen (e.g. initializing DRM fails) and we don't assert (for a good reason).

>when we are working with sockets/pipes (not something on NFS) close can not fail...

I don't see this written in POSIX.

Reply:
Personally I don't like asserts and never use them in my own code so it was used just as an example.

> I don't see this written in POSIX.

Not everything is written in POSIX, check this one please http://austingroupbugs.net/view.php?id=529
It is about correcting description of posix close and adding restart variant.
Current state of close is that even if it fail you can not do anything with it.
On Linux it strictly can not be interrupted (EINTR) since it can not be restarted.
So what is point here to fail? What can you do? Restart close?

Reply:
Look, this is simple: `close` can fail, thus check for errors, and fail if there's one.

Sorry, I'm getting a little tired of this discussion.

Reply:
Have you checked link to austingroup and discussion there regarding failing close?

Reply:
Also let me clarify my point:

close(2), even failed, closes descritpor. It's not stated in current POSIX but will be added into next revision.

All we want from close(2) here is to close fd and that's done even if (theoretical) error returned.

For us that error means nothing - fd is already closed, no leaking, correct cleanup.

Reply:
>For checking close(2) return value I've added small wrapper close_warn that logs error

-1

>For us that error means nothing - fd is already closed, no leaking, correct cleanup.

No - it means FD is not a valid file descriptor, or something else went wrong. Either way, something bad happen. Let's not ignore it.

Reply:
There are 2 places with call to close: cleanup in error path and closing client side of socket pair/pipe.

Cleanup path should not be interrupted - it can lead to fd leaking when we fail on first close.
We can log errors there but can not return in the middle.

Closing client side of socketpair/pipe is done after client is spawned. If we want to fail there we have to kill client process. Otherwise we report error to caller but wayland client is already started. But it's hard to kill client process - thanks to double fork we have no control over it. Looks bad to me.

Also brief overview of close errors for BSD/Linux:

EBADF: fd is not valid? When it was closed? There are ~10 lines of code between socketpair and close.
All other errors (EINTR, ECONNRESET(BSD), ENOSPC, EDQUOT, EIO(mystical Linux)) are just informational, fd is closed. So why should we hard fail?

https://www.freebsd.org/cgi/man.cgi?query=close&sektion=2
http://man7.org/linux/man-pages/man2/close.2.html

Reply:
*scratches head* slightly confused but the disconnect going on here.

the original code being replaced by this function was just warning anyways. the assertion that if an error occurs then fail doesn't always hold. sometimes just logging and moving on is the correct answer to handle an error.

there is a potential behavior change here because of one spot where the code doesn't return early, but @shramov has covered that in their explanation.

basically looks like everyone is bikeshedding over a non-issue.

@emersion its likely not a real issue as @shramov pointed out and provided documentation and reasonable explanations for.

@shramov you could just return a success bool and short circuit the calling function to maintain previous behavior and if you really feel strongly over not hard failing open another PR around that particular issue.

apologies for just randomly dropping my 2c.